File: app/src/main/java/com/asfoundation/wallet/transactions/TransactionsMapper.java
package com.asfoundation.wallet.transactions;

import com.asfoundation.wallet.entity.WalletHistory;
import java.util.ArrayList;
import java.util.List;
import org.jetbrains.annotations.NotNull;

import static com.asfoundation.wallet.transactions.Transaction.TransactionType.ADS_OFFCHAIN;
import static com.asfoundation.wallet.transactions.Transaction.TransactionType.BONUS;
import static com.asfoundation.wallet.transactions.Transaction.TransactionType.ETHER_TRANSFER;
import static com.asfoundation.wallet.transactions.Transaction.TransactionType.IAP;
import static com.asfoundation.wallet.transactions.Transaction.TransactionType.IAP_OFFCHAIN;
import static com.asfoundation.wallet.transactions.Transaction.TransactionType.STANDARD;
import static com.asfoundation.wallet.transactions.Transaction.TransactionType.TOP_UP;
import static com.asfoundation.wallet.transactions.Transaction.TransactionType.TRANSFER_OFF_CHAIN;

public class TransactionsMapper {

  private static final String PERK_BONUS = "perk_bonus";
  private static final String GAMIFICATION_LEVEL_UP = "GAMIFICATION_LEVEL_UP";
  private static final String PACKAGE_PERK = "PACKAGE_PERK";

  public TransactionsMapper() {
  }

  public List<Transaction> mapTransactionsFromWalletHistory(
      List<WalletHistory.Transaction> transactions) {
    List<Transaction> transactionList = new ArrayList<>(transactions.size());
    for (int i = transactions.size() - 1; i >= 0; i--) {
      WalletHistory.Transaction transaction = transactions.get(i);

      Transaction.TransactionType txType = mapTransactionType(transaction);

      Transaction.TransactionStatus status;
      switch (transaction.getStatus()) {
        case SUCCESS:
          status = Transaction.TransactionStatus.SUCCESS;
          break;
        default:
        case FAIL:
          status = Transaction.TransactionStatus.FAILED;
          break;
      }

      String sourceName;
      Transaction.SubType bonusSubType;
      Transaction.Perk perk;
      if (txType.equals(BONUS)) {
        if (transaction.getBonus() == null) {
          sourceName = null;
        } else {
          sourceName = transaction.getBonus()
              .stripTrailingZeros()
              .toPlainString();
        }
      } else {
        sourceName = transaction.getApp();
      }
      bonusSubType = mapSubtype(transaction.getSubType());
      perk = mapPerk(transaction.getPerk());
      transactionList.add(0,
          new Transaction(transaction.getTxID(), txType, bonusSubType, transaction.getTitle(),
              transaction.getDescription(), perk, null, transaction.getTs()
              .getTime(), transaction.getProcessedTime()
              .getTime(), status, transaction.getAmount()
              .toString(), transaction.getSender(), transaction.getReceiver(),
              new TransactionDetails(sourceName,
                  new TransactionDetails.Icon(TransactionDetails.Icon.Type.URL,
                      transaction.getIcon()), transaction.getSku()),
              txType.equals(ETHER_TRANSFER) ? "ETH" : "APPC",
              mapOperations(transaction.getOperations())));
    }
    return transactionList;
  }

  private Transaction.Perk mapPerk(String perk) {
    Transaction.Perk perkType = null;
    if (perk != null) {
      switch (perk) {
        case GAMIFICATION_LEVEL_UP:
          perkType = Transaction.Perk.GAMIFICATION_LEVEL_UP;
          break;
        case PACKAGE_PERK:
          perkType = Transaction.Perk.PACKAGE_PERK;
          break;
        default:
          perkType = Transaction.Perk.UNKNOWN;
          break;
      }
    }
    return perkType;
  }

  private Transaction.SubType mapSubtype(String subType) {
    Transaction.SubType bonusSubType = null;
    if (subType != null) {
      if (subType.equals(PERK_BONUS)) {
        bonusSubType = Transaction.SubType.PERK_PROMOTION;
      } else {
        bonusSubType = Transaction.SubType.UNKNOWN;
      }
    }
    return bonusSubType;
  }

  private List<Operation> mapOperations(List<WalletHistory.Operation> operations) {
    List<Operation> list = new ArrayList<>(operations.size());

    for (WalletHistory.Operation operation : operations) {
      list.add(new Operation(operation.getTransactionId(), operation.getSender(),
          operation.getReceiver(), operation.getFee()));
    }
    return list;
  }

  @NotNull
  private Transaction.TransactionType mapTransactionType(WalletHistory.Transaction transaction) {
    switch (transaction.getType()) {
      case "Transfer OffChain":
        return TRANSFER_OFF_CHAIN;
      case "Topup OffChain":
        return TOP_UP;
      case "IAP OffChain":
        return IAP_OFFCHAIN;
      case "bonus":
        return BONUS;
      case "PoA OffChain":
        return ADS_OFFCHAIN;
      case "Ether Transfer":
        return ETHER_TRANSFER;
      case "IAP":
        return IAP;
      default:
        return STANDARD;
    }
  }
}

File: app/src/main/java/com/asfoundation/wallet/transactions/TransactionsMapper.kt
package com.asfoundation.wallet.transactions

import com.appcoins.wallet.core.network.backend.model.WalletHistory
import com.asfoundation.wallet.repository.entity.OperationEntity
import com.asfoundation.wallet.repository.entity.TransactionDetailsEntity
import com.asfoundation.wallet.repository.entity.TransactionEntity
import java.util.*
import javax.inject.Inject

class TransactionsMapper @Inject constructor() {

  companion object {
    private const val PERK_BONUS = "perk_bonus"
    private const val GAMIFICATION_LEVEL_UP = "GAMIFICATION_LEVEL_UP"
    private const val PACKAGE_PERK = "PACKAGE_PERK"
  }

  fun map(transaction: WalletHistory.Transaction, wallet: String): TransactionEntity {
    val txType = mapTransactionType(transaction)
    val method = mapTransactionMethod(transaction.method)
    val status = map(transaction.status)
    val sourceName = mapSource(txType, transaction)
    val bonusSubType = mapSubtype(transaction.subType)
    val perk = mapPerk(transaction.perk)
    val operations = mapOperations(transaction.operations)
    val icon = TransactionDetailsEntity.Icon(TransactionDetailsEntity.Type.URL, transaction.icon)
    val details = TransactionDetailsEntity(icon, sourceName, transaction.sku)
    val currency = if (txType == TransactionEntity.TransactionType.ETHER_TRANSFER) "ETH" else "APPC"
    return TransactionEntity(transaction.txID, wallet, null, perk, txType, method, bonusSubType,
        transaction.title, transaction.description, transaction.ts.time,
        transaction.processedTime.time, status, transaction.amount.toString(), currency,
        transaction.paidAmount, transaction.paidCurrency, transaction.sender,
        transaction.receiver, details, operations, transaction.orderReference)
  }

  private fun mapSource(txType: TransactionEntity.TransactionType,
                        transaction: WalletHistory.Transaction): String? {
    return if (txType == TransactionEntity.TransactionType.BONUS) {
      if (transaction.bonus == null) {
        null
      } else {
        transaction.bonus.stripTrailingZeros()
            .toPlainString()
      }
    } else {
      transaction.app
    }
  }

  private fun map(status: WalletHistory.Status): TransactionEntity.TransactionStatus {
    return when (status) {
      WalletHistory.Status.SUCCESS -> TransactionEntity.TransactionStatus.SUCCESS
      WalletHistory.Status.FAIL -> TransactionEntity.TransactionStatus.FAILED
    }
  }

  fun isRevertType(type: String): Boolean {
    return when (type) {
      "Bonus Revert OffChain",
      "Topup Revert OffChain",
      "IAP Revert OffChain" -> true
      else -> false
    }
  }

  private fun mapTransactionType(
      transaction: WalletHistory.Transaction): TransactionEntity.TransactionType {
    return when (transaction.type) {
      "Transfer" -> TransactionEntity.TransactionType.TRANSFER
      "Transfer OffChain" -> TransactionEntity.TransactionType.TRANSFER_OFF_CHAIN
      "Ether Transfer" -> TransactionEntity.TransactionType.ETHER_TRANSFER
      "Topup OffChain" -> TransactionEntity.TransactionType.TOP_UP
      "IAP OffChain" -> TransactionEntity.TransactionType.IAP_OFFCHAIN
      "bonus" -> TransactionEntity.TransactionType.BONUS
      "PoA OffChain" -> TransactionEntity.TransactionType.ADS_OFFCHAIN
      "IAP" -> TransactionEntity.TransactionType.IAP
      "Bonus Revert OffChain" -> TransactionEntity.TransactionType.BONUS_REVERT
      "Topup Revert OffChain" -> TransactionEntity.TransactionType.TOP_UP_REVERT
      "IAP Revert OffChain" -> TransactionEntity.TransactionType.IAP_REVERT
      "Subscription OffChain" -> TransactionEntity.TransactionType.INAPP_SUBSCRIPTION
      "Eskills Reward OffChain" -> TransactionEntity.TransactionType.ESKILLS_REWARD
      "Eskills OffChain" -> TransactionEntity.TransactionType.ESKILLS
      else -> TransactionEntity.TransactionType.STANDARD
    }
  }

  private fun mapTransactionMethod(method: String?): TransactionEntity.Method {
    return when (method) {
      "appcoins_credits" -> TransactionEntity.Method.APPC_C
      "appcoins" -> TransactionEntity.Method.APPC
      "ETH" -> TransactionEntity.Method.ETH
      else -> TransactionEntity.Method.UNKNOWN
    }
  }

  private fun mapPerk(perk: String?): TransactionEntity.Perk? {
    return perk?.let {
      when (it) {
        GAMIFICATION_LEVEL_UP -> TransactionEntity.Perk.GAMIFICATION_LEVEL_UP
        PACKAGE_PERK -> TransactionEntity.Perk.PACKAGE_PERK
        else -> TransactionEntity.Perk.UNKNOWN
      }
    }
  }

  private fun mapSubtype(subType: String?): TransactionEntity.SubType? {
    var bonusSubType: TransactionEntity.SubType? = null
    if (subType != null) {
      bonusSubType = if (subType == PERK_BONUS) {
        TransactionEntity.SubType.PERK_PROMOTION
      } else {
        TransactionEntity.SubType.UNKNOWN
      }
    }
    return bonusSubType
  }

  private fun mapOperations(operations: List<WalletHistory.Operation>): List<OperationEntity> {
    val list: MutableList<OperationEntity> = ArrayList(operations.size)
    for (operation in operations) {
      list.add(OperationEntity(operation.transactionId,
          operation.sender,
          operation.receiver, operation.fee))
    }
    return list
  }

}